docker/sds-api/spine-directory-service/sds/request/routing_reliability_handler.py

```diff
@@ -42,23 +55,83 @@ async def get(self):
+            fr_service_ids = self._get_forward_reliable_service_ids(ldap_results)
+            fr_addresses_map = await self._get_forward_reliable_addresses_map(fr_service_ids)
+            self._replace_forward_reliable_endpoint(ldap_results, fr_addresses_map)
+
+    def _get_forward_reliable_service_ids(self, ldap_results: List[dict]) -> Set[str]:
+        service_ids = map(lambda single_result: single_result['nhsMhsSvcIA'], ldap_results)
+        only_fr_service_ids = filter(lambda service_id: self._is_forward_reliable_service(service_id), service_ids)
+        return set(only_fr_service_ids)
+
+    async def _get_forward_reliable_addresses_map(self, fr_service_ids: Set[str]) -> Dict:
+        result = {}
+        for fr_service_id in fr_service_ids:
+            fr_address = await self._get_forward_reliable_address(fr_service_id)
+            result[fr_service_id] = fr_address
+        return result
+
+    @staticmethod
+    def _is_spine_core_ods_code(org_code):
+        spine_core_ods_code = config.get_config('SPINE_CORE_ODS_CODE')
+        return spine_core_ods_code == org_code
+
+    async def _get_forward_reliable_address(self, service_id: str) -> str:
+        spine_core_ods_code = config.get_config('SPINE_CORE_ODS_CODE')
+        logger.info("Looking up forward reliable routing and reliability information. {org_code}, {service_id}",
+                    fparams={"org_code": spine_core_ods_code, "service_id": service_id})
+        ldap_results = await self.sds_client.get_mhs_details(spine_core_ods_code, service_id)
+        logger.info("Obtained forward reliable routing and reliability information. {ldap_results}",
+                    fparams={"ldap_results": ldap_results})
+
+        if len(ldap_results) != 1:
+            raise ValueError(f"Expected 1 result for forward reliable routing and reliability but got {str(len(ldap_results))}")
+
+        addresses = ldap_results[0]['nhsMHSEndPoint']
+
+        if len(addresses) != 1:
+            raise ValueError(f"Expected 1 address for forward reliable routing and reliability but got {str(len(addresses))}")
+
+        return addresses[0]
+
+    def _replace_forward_reliable_endpoint(self, ldap_results: List[dict], fr_addresses_map: Dict[str, str]):
+        fr_ldap_result = (ldap_result for ldap_result in ldap_results if self._is_forward_reliable_service(ldap_result['nhsMhsSvcIA']))
+        for single_fr_ldap_result in fr_ldap_result:
+            single_fr_ldap_result['nhsMHSEndPoint'] = [fr_addresses_map[single_fr_ldap_result['nhsMhsSvcIA']]]
+
+    @staticmethod
+    def _is_forward_reliable_service(service_interaction: str):
+        if not service_interaction:
+            return False
+
+        parts = service_interaction.split(':')
+
+        if len(parts) < 2:
+            raise RuntimeError(f"Invalid service interaction {service_interaction}")
+
+        return parts[-2] in FORWARD_RELIABLE_SERVICES and parts[-1] in FORWARD_RELIABLE_INTERACTIONS
```

Comment:
Maybe it would be worth to set variables for `parts` indexes `-2` and `-1` describing what each of them is?